File: apollo-runtime/src/main/java/com/apollographql/apollo/response/CustomTypeValue.java
package com.apollographql.apollo.response;

import com.apollographql.apollo.internal.json.Utils;

import java.util.List;
import java.util.Map;

/**
 * A wrapper class for representation of custom GraphQL type value, used in user provided {@link CustomTypeAdapter}
 * encoding / decoding functions.
 *
 * @param <T> value type, one of the supported types <ul><li>{@link java.lang.String} <li>{@link
 *            java.lang.Boolean}<li>{@link java.lang.Number}<li>{@link Map}<li>{@link List}<ul/>
 */
public abstract class CustomTypeValue<T> {
  public final T value;

  public static CustomTypeValue fromRawValue(Object value) {
    if (value instanceof Map || value instanceof List) {
      try {
        return new GraphQLJsonString(Utils.toJsonString(value));
      } catch (Exception e) {
        throw new RuntimeException(e);
      }
    } else if (value instanceof Boolean) {
      return new GraphQLBoolean((Boolean) value);
    } else if (value instanceof Number) {
      return new GraphQLNumber((Number) value);
    } else {
      return new GraphQLString(value.toString());
    }
  }

  private CustomTypeValue(T value) {
    this.value = value;
  }

  /**
   * Represents a {@code String} value
   */
  public static class GraphQLString extends CustomTypeValue<String> {
    public GraphQLString(String value) {
      super(value);
    }
  }

  /**
   * Represents a {@code Boolean} value
   */
  public static class GraphQLBoolean extends CustomTypeValue<Boolean> {
    public GraphQLBoolean(Boolean value) {
      super(value);
    }
  }

  /**
   * Represents a {@code Number} value
   */
  public static class GraphQLNumber extends CustomTypeValue<Number> {
    public GraphQLNumber(Number value) {
      super(value);
    }
  }

  /**
   * Represents a {@code JsonString} value
   */
  public static class GraphQLJsonString extends CustomTypeValue<String> {
    public GraphQLJsonString(String value) {
      super(value);
    }
  }

  /**
   * Represents a JSON object value, will serialized as an regular json object
   */
  public static class GraphQLJson extends CustomTypeValue<Map<String, Object>> {
    public GraphQLJson(Map<String, Object> value) {
      super(value);
    }
  }
}

File: apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/CustomTypeValue.kt
package com.apollographql.apollo.api

import kotlin.jvm.JvmField
import kotlin.jvm.JvmStatic

/**
 * A wrapper class for representation of custom GraphQL type value, used in user provided [CustomScalarTypeAdapter]
 * encoding / decoding functions.
 **/
sealed class CustomTypeValue<T>(@JvmField val value: T) {

  object GraphQLNull: CustomTypeValue<Unit>(Unit)

  /**
   * Represents a `String` value
   */
  class GraphQLString(value: String) : CustomTypeValue<String>(value) {
    override fun equals(other: Any?): Boolean {
      if (this === other) return true
      if (other !is GraphQLString) return false
      if (value != other.value) return false
      return true
    }

    override fun hashCode(): Int {
      return value.hashCode()
    }
  }

  /**
   * Represents a `Boolean` value
   */
  class GraphQLBoolean(value: Boolean) : CustomTypeValue<Boolean>(value) {
    override fun equals(other: Any?): Boolean {
      if (this === other) return true
      if (other !is GraphQLBoolean) return false
      if (value != other.value) return false
      return true
    }

    override fun hashCode(): Int {
      return value.hashCode()
    }
  }

  /**
   * Represents a `Number` value
   */
  class GraphQLNumber(value: Number) : CustomTypeValue<Number>(value) {
    override fun equals(other: Any?): Boolean {
      if (this === other) return true
      if (other !is GraphQLNumber) return false
      if (value != other.value) return false
      return true
    }

    override fun hashCode(): Int {
      return value.hashCode()
    }
  }

  /**
   * Represents a JSON `Object` value
   */
  class GraphQLJsonObject(value: Map<String, Any>) : CustomTypeValue<Map<String, Any>>(value) {
    override fun equals(other: Any?): Boolean {
      if (this === other) return true
      if (other !is GraphQLJsonObject) return false
      if (value != other.value) return false
      return true
    }

    override fun hashCode(): Int {
      return value.hashCode()
    }
  }

  /**
   * Represents a JSON `List` value
   */
  class GraphQLJsonList(value: List<Any>) : CustomTypeValue<List<Any>>(value) {
    override fun equals(other: Any?): Boolean {
      if (this === other) return true
      if (other !is GraphQLJsonList) return false
      if (value != other.value) return false
      return true
    }

    override fun hashCode(): Int {
      return value.hashCode()
    }
  }

  companion object {

    // TODO: should this method accept nullable values in case a scalar type is represented as 'null'?
    @JvmStatic
    @Suppress("UNCHECKED_CAST")
    fun fromRawValue(value: Any): CustomTypeValue<*> {
      return when (value) {
        is Map<*, *> -> GraphQLJsonObject(value as Map<String, Any>)
        is List<*> -> GraphQLJsonList(value as List<Any>)
        is Boolean -> GraphQLBoolean(value)
        is BigDecimal -> GraphQLNumber(value.toNumber())
        is Number -> GraphQLNumber(value)
        else -> GraphQLString(value.toString())
      }
    }
  }
}
